eppo_core/src/attributes.rs

```diff
@@ -58,43 +24,177 @@ impl Subject {
 /// ```
 pub type Attributes = HashMap<String, AttributeValue>;
 
-/// Enum representing possible values of an attribute for a subject.
+/// Attribute of a subject or action.
 ///
-/// Conveniently implements `From` conversions for `String`, `&str`, `f64`, and `bool` types.
+/// Stores attribute value (string, number, boolean) along with attribute kind (numeric or
+/// categorical). Storing kind is helpful to make `Attributes` ↔ `ContextAttributes` conversion
+/// isomorphic.
 ///
-/// Examples:
-/// ```
-/// # use eppo_core::AttributeValue;
-/// let string_attr: AttributeValue = "example".into();
-/// let number_attr: AttributeValue = 42.0.into();
-/// let bool_attr: AttributeValue = true.into();
-/// ```
-#[derive(Debug, Serialize, Deserialize, PartialEq, PartialOrd, From, Clone)]
+/// Note that attribute kind is stripped during serialization, so Attribute → JSON → Attribute
+/// conversion is lossy.
+#[derive(Debug, Clone, PartialEq, PartialOrd, derive_more::From, Serialize, Deserialize)]
+#[from(NumericAttribute, CategoricalAttribute, f64, bool, Str, String, &str, Arc<str>, Arc<String>, Cow<'_, str>)]
+pub struct AttributeValue(AttributeValueImpl);
+#[derive(Debug, Clone, PartialEq, PartialOrd, derive_more::From, Serialize, Deserialize)]
 #[serde(untagged)]
-pub enum AttributeValue {
-    /// A string value.
+enum AttributeValueImpl {
+    #[from(NumericAttribute, f64)]
+    Numeric(NumericAttribute),
+    #[from(CategoricalAttribute, Str, bool, String, &str, Arc<str>, Arc<String>, Cow<'_, str>)]
+    Categorical(CategoricalAttribute),
     #[from(ignore)]
+    Null,
+}
```

Comment:
All enum variants have the same visibility as enum itself, so anyone who had access to `AttributeValue` was able to create or pattern-match on it.

I wanted to hide our implementation, so splitting into public struct (with private constructor) and a private enum is a common trick in Rust